subprojects/testing-native/src/main/java/org/gradle/nativeplatform/test/xctest/plugins/XCTestConventionPlugin.java

```diff
@@ -177,15 +177,13 @@ public void execute(SwiftLibraryPlugin plugin) {
 
     private void configureTestedSwiftComponent(Project project) {
         TaskContainer tasks = project.getTasks();
-        DirectoryVar buildDirectory = project.getLayout().getBuildDirectory();
-
-        SwiftCompile compile = tasks.withType(SwiftCompile.class).getByName("compileSwift");
 
+        SwiftCompile compileMain = tasks.withType(SwiftCompile.class).getByName("compileSwift");
         SwiftCompile compileTest = tasks.withType(SwiftCompile.class).getByName("compileTestSwift");
-        compileTest.dependsOn(compile);
-        compileTest.includes(buildDirectory.dir("main/objs"));
+        compileTest.includes(compileMain.getObjectFileDirectory());
 
+        AbstractLinkTask linkMain = tasks.withType(AbstractLinkTask.class).getByName("linkMain");
         AbstractLinkTask linkTest = tasks.withType(AbstractLinkTask.class).getByName("linkTest");
-        linkTest.source(compile.getObjectFileDirectory().getAsFileTree().matching(new PatternSet().include("**/*.obj", "**/*.o").exclude("**/main.o")));
+        linkTest.source(linkMain.getSource());
```

Comment:
Although I'm sure it works, this smells wrong to me.  We should be setting the inputs of this link task to something that's an output of another task - it shouldn't be the inputs of some similar task (even if those inputs happen to be correct).  I would leave this as the compile task objectFileDirectory like it was before.

In your test, the problem is that `AbstractLinkTask.setSource()` does not set the file collection - it calls `setFrom()` on the file collection and changes the contents.  So you're never going to be able to call `setSource()` with a filtered view of itself.  You can only 1) add an additional thing to the contents or 2) set the contents to something different.  You're going to need to set it to a new `FileCollection` object and that object will also need to be a buildable thing (to drive the proper task dependencies).  In other words, you want to declare it to something more concrete than "whatever it already is, just not main.o", such as "the buildable object file directory from the compile task that only includes .o file and .obj files but not main.o" (i.e. `compile.getObjectFileDirectory()...`).

The other option would be to change `AbstractLinkTask` to behave more like `SourceTask` and maintain the list of source inputs and a configurable list of filters and only construct the `FileCollection` when it is requested.  In fact, I wonder if `AbstractLinkTask` could just _be_ a `SourceTask`...